File: packages/kotlin/sdk/src/main/kotlin/dev/usermon/sdk/UsermonClient.kt
package dev.usermon.sdk

import android.content.Context
import android.os.Build
import kotlinx.coroutines.*
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.security.SecureRandom
import java.util.concurrent.CopyOnWriteArrayList

// ── Configuration ─────────────────────────────────────────────────────────────

data class UsermonConfig(
    /** Full ingest base URL, e.g. https://xxx.convex.site */
    val ingestUrl: String,
    /** Project ingest key (um_...) */
    val ingestKey: String,
    /** Default platform (default: ANDROID) */
    val platform: Platform = Platform.ANDROID,
    /** Default release/version */
    val release: String? = null,
    /** Flush interval in ms (default 5000) */
    val flushIntervalMs: Long = 5_000L,
    /** Max queue size before forced flush (default 50) */
    val maxBatchSize: Int = 50,
    /** HTTP timeout in ms (default 10000) */
    val timeoutMs: Int = 10_000,
)

// ── Client ────────────────────────────────────────────────────────────────────

/**
 * Thread-safe Usermon client for Android.
 *
 * ```kotlin
 * // Application.onCreate():
 * Usermon.configure(
 *     context = this,
 *     ingestUrl = "https://xxx.convex.site",
 *     ingestKey = "um_xxx",
 *     release = BuildConfig.VERSION_NAME,
 * )
 *
 * // Anywhere:
 * Usermon.captureException("Something broke", route = "/screen/home")
 * Usermon.captureLog("User checked out", level = LogLevel.INFO, attrs = mapOf("cart_size" to 3))
 * ```
 */
class UsermonClient(private val config: UsermonConfig) {

    /** The session key for this app session. */
    val sessionKey: String = newSessionKey()

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val queuedSessions = CopyOnWriteArrayList<Session>()
    private val queuedRumEvents = CopyOnWriteArrayList<RumEvent>()
    private val queuedApiSpans = CopyOnWriteArrayList<ApiSpan>()
    private val queuedExceptions = CopyOnWriteArrayList<UsermonException>()
    private val queuedLogs = CopyOnWriteArrayList<LogEvent>()

    init {
        startFlushTimer()
    }

    // ── Capture API ───────────────────────────────────────────────────────────

    fun captureException(
        message: String,
        stack: String? = null,
        route: String? = null,
        level: IssueLevel = IssueLevel.ERROR,
        traceId: String? = null,
        release: String? = null,
    ) {
        queuedExceptions.add(
            UsermonException(
                message = message,
                stack = stack,
                platform = config.platform,
                route = route,
                level = level,
                sessionKey = sessionKey,
                traceId = traceId,
                release = release ?: config.release,
            )
        )
        flushIfNeeded()
    }

    fun capture(throwable: Throwable, route: String? = null, traceId: String? = null) {
        captureException(
            message = throwable.message ?: throwable.javaClass.simpleName,
            stack = throwable.stackTraceToString(),
            route = route,
            traceId = traceId,
        )
    }

    fun captureLog(
        message: String,
        level: LogLevel = LogLevel.INFO,
        attrs: Map<String, Any>? = null,
        traceId: String? = null,
        release: String? = null,
    ) {
        val attrsJson = attrs?.let { JSONObject(it).toString() }
        queuedLogs.add(
            LogEvent(
                level = level,
                message = message,
                attrsJson = attrsJson,
                platform = config.platform,
                sessionKey = sessionKey,
                traceId = traceId,
                release = release ?: config.release,
            )
        )
        flushIfNeeded()
    }

    fun captureSpan(
        method: String,
        route: String,
        status: Int,
        durationMs: Long,
        traceId: String? = null,
        spanId: String? = null,
        release: String? = null,
    ) {
        queuedApiSpans.add(
            ApiSpan(
                method = method.uppercase(),
                route = route,
                status = status,
                durationMs = durationMs,
                platform = config.platform,
                traceId = traceId,
                spanId = spanId,
                sessionKey = sessionKey,
                release = release ?: config.release,
            )
        )
        flushIfNeeded()
    }

    fun captureRumEvent(
        type: RumEventType,
        name: String,
        value: Double? = null,
        route: String? = null,
        release: String? = null,
    ) {
        queuedRumEvents.add(
            RumEvent(
                type = type,
                name = name,
                value = value,
                platform = config.platform,
                route = route,
                sessionKey = sessionKey,
                release = release ?: config.release,
            )
        )
        flushIfNeeded()
    }

    /** Convenience: capture an Android screen view. */
    fun captureScreenView(screenName: String) {
        captureRumEvent(type = RumEventType.SCREEN_VIEW, name = screenName, route = screenName)
    }

    // ── Session ───────────────────────────────────────────────────────────────

    fun startSession(deviceOs: String? = null, release: String? = null): String {
        val os = deviceOs ?: "Android ${Build.VERSION.RELEASE}"
        queuedSessions.add(
            Session(
                sessionKey = sessionKey,
                platform = config.platform,
                release = release ?: config.release,
                deviceOs = os,
            )
        )
        flushIfNeeded()
        return sessionKey
    }

    // ── Direct send ───────────────────────────────────────────────────────────

    suspend fun health(): JSONObject = withContext(Dispatchers.IO) {
        val url = URL("${config.ingestUrl.trimEnd('/')}/health")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = config.timeoutMs
            readTimeout = config.timeoutMs
        }
        val body = conn.inputStream.bufferedReader().readText()
        JSONObject(body)
    }

    suspend fun ingest(batch: IngestBatch): JSONObject = withContext(Dispatchers.IO) {
        val url = URL("${config.ingestUrl.trimEnd('/')}/v1/ingest")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            setRequestProperty("Authorization", "Bearer ${config.ingestKey}")
            setRequestProperty("Content-Type", "application/json")
            doOutput = true
            connectTimeout = config.timeoutMs
            readTimeout = config.timeoutMs
        }
        val payload = batch.toJson().toString()
        OutputStreamWriter(conn.outputStream).use { it.write(payload) }
        val code = conn.responseCode
        val body = if (code < 400) conn.inputStream.bufferedReader().readText()
                   else conn.errorStream?.bufferedReader()?.readText() ?: ""
        if (code >= 400) error("Usermon ingest error $code: $body")
        JSONObject(body)
    }

    // ── Flush ─────────────────────────────────────────────────────────────────

    fun flush() {
        scope.launch { flushSuspend() }
    }

    suspend fun flushSuspend() {
        val batch = IngestBatch(
            sessions = queuedSessions.toList().also { queuedSessions.clear() },
            rumEvents = queuedRumEvents.toList().also { queuedRumEvents.clear() },
            apiSpans = queuedApiSpans.toList().also { queuedApiSpans.clear() },
            exceptions = queuedExceptions.toList().also { queuedExceptions.clear() },
            logs = queuedLogs.toList().also { queuedLogs.clear() },
        )
        if (batch.isEmpty()) return
        try {
            ingest(batch)
        } catch (e: Exception) {
            android.util.Log.w("Usermon", "flush failed: ${e.message}")
        }
    }

    // ── Private ───────────────────────────────────────────────────────────────

    private fun flushIfNeeded() {
        val total = queuedSessions.size + queuedRumEvents.size + queuedApiSpans.size +
                    queuedExceptions.size + queuedLogs.size
        if (total >= config.maxBatchSize) flush()
    }

    private fun startFlushTimer() {
        scope.launch {
            while (isActive) {
                delay(config.flushIntervalMs)
                flushSuspend()
            }
        }
    }

    private fun newSessionKey(): String {
        val bytes = ByteArray(16)
        SecureRandom().nextBytes(bytes)
        return bytes.joinToString("") { "%02x".format(it) }
    }

    fun close() {
        scope.launch { flushSuspend() }
        scope.cancel()
    }
}

// ── Singleton ─────────────────────────────────────────────────────────────────

object Usermon {
    @Volatile private var _client: UsermonClient? = null

    /**
     * Configure the global Usermon client.
     * Call once in `Application.onCreate()`.
     */
    fun configure(
        context: Context,
        ingestUrl: String,
        ingestKey: String,
        platform: Platform = Platform.ANDROID,
        release: String? = null,
        flushIntervalMs: Long = 5_000L,
    ): UsermonClient {
        val versionName = try {
            context.packageManager.getPackageInfo(context.packageName, 0).versionName
        } catch (_: Exception) { null }

        val client = UsermonClient(
            UsermonConfig(
                ingestUrl = ingestUrl,
                ingestKey = ingestKey,
                platform = platform,
                release = release ?: versionName,
                flushIntervalMs = flushIntervalMs,
            )
        )
        _client = client
        client.startSession()
        return client
    }

    /** The configured global client. Throws if `configure()` was not called. */
    val shared: UsermonClient
        get() = _client ?: error("[Usermon] Call Usermon.configure() before using Usermon.shared")

    /** Returns the global client, or null if not yet configured. */
    val current: UsermonClient? get() = _client

    // Forwarding helpers
    fun captureException(message: String, stack: String? = null, route: String? = null, level: IssueLevel = IssueLevel.ERROR) =
        _client?.captureException(message, stack, route, level)

    fun capture(throwable: Throwable, route: String? = null) =
        _client?.capture(throwable, route)

    fun captureLog(message: String, level: LogLevel = LogLevel.INFO, attrs: Map<String, Any>? = null) =
        _client?.captureLog(message, level, attrs)

    fun captureSpan(method: String, route: String, status: Int, durationMs: Long) =
        _client?.captureSpan(method, route, status, durationMs)

    fun captureScreenView(screenName: String) =
        _client?.captureScreenView(screenName)

    fun flush() = _client?.flush()
}

File: packages/swift/Sources/UsermonSDK/UsermonClient.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

// MARK: - Errors

public enum UsermonError: Error, LocalizedError {
    case invalidURL
    case httpError(statusCode: Int, body: String)
    case encodingError(Error)
    case decodingError(Error)
    case networkError(Error)

    public var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid ingest URL"
        case .httpError(let code, let body): return "HTTP \(code): \(body)"
        case .encodingError(let e): return "Encoding error: \(e)"
        case .decodingError(let e): return "Decoding error: \(e)"
        case .networkError(let e): return "Network error: \(e)"
        }
    }
}

// MARK: - Configuration

public struct UsermonConfiguration: Sendable {
    /// Full ingest base URL, e.g. https://xxx.convex.site
    public var ingestUrl: String
    /// Project ingest key (um_...)
    public var ingestKey: String
    /// Default platform (default: .ios)
    public var platform: Platform
    /// App release/version string (default: CFBundleShortVersionString)
    public var release: String?
    /// Sample rate 0–1 for auto-instrumented URLSession spans. Default 1.
    public var tracesSampleRate: Double
    /// Flush interval in seconds. Default 5.
    public var flushInterval: TimeInterval
    /// Maximum batch size before forced flush. Default 50.
    public var maxBatchSize: Int

    public init(
        ingestUrl: String,
        ingestKey: String,
        platform: Platform = .ios,
        release: String? = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String,
        tracesSampleRate: Double = 1.0,
        flushInterval: TimeInterval = 5.0,
        maxBatchSize: Int = 50
    ) {
        self.ingestUrl = ingestUrl
        self.ingestKey = ingestKey
        self.platform = platform
        self.release = release
        self.tracesSampleRate = tracesSampleRate
        self.flushInterval = flushInterval
        self.maxBatchSize = maxBatchSize
    }
}

// MARK: - Client

/// Thread-safe Usermon client for iOS/macOS.
///
/// ```swift
/// import UsermonSDK
///
/// // In AppDelegate or @main:
/// Usermon.configure(
///     ingestUrl: "https://xxx.convex.site",
///     ingestKey: "um_xxx"
/// )
///
/// // Anywhere:
/// Usermon.shared.captureException(message: "Oops", route: "/screen/home")
/// Usermon.shared.captureLog("User tapped button", level: .info, attrs: ["button": "checkout"])
/// ```
public final class UsermonClient: @unchecked Sendable {

    private let config: UsermonConfiguration
    private let urlSession: URLSession
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    // Queue
    private let lock = NSLock()
    private var queuedSessions: [Session] = []
    private var queuedRumEvents: [RumEvent] = []
    private var queuedApiSpans: [ApiSpan] = []
    private var queuedExceptions: [UsermonException] = []
    private var queuedLogs: [LogEvent] = []

    private var flushTimer: Timer?

    /// The session key for the current app session.
    public let sessionKey: String

    public init(configuration: UsermonConfiguration) {
        self.config = configuration
        self.sessionKey = hexId(bytes: 16)
        self.urlSession = URLSession(configuration: .default)
        self.encoder = JSONEncoder()
        self.decoder = JSONDecoder()

        // Auto-detect device OS
        startFlushTimer()
    }

    deinit {
        flushTimer?.invalidate()
    }

    // MARK: - Capture API

    /// Capture an exception and buffer it for the next flush.
    public func captureException(
        message: String,
        stack: String? = nil,
        route: String? = nil,
        level: IssueLevel = .error,
        traceId: String? = nil,
        release: String? = nil
    ) {
        let exc = UsermonException(
            message: message,
            stack: stack,
            platform: config.platform,
            route: route,
            level: level,
            sessionKey: sessionKey,
            traceId: traceId,
            release: release ?? config.release
        )
        lock.lock(); queuedExceptions.append(exc); lock.unlock()
        flushIfNeeded()
    }

    /// Capture an `Error` object.
    public func capture(error: Error, route: String? = nil, traceId: String? = nil) {
        captureException(
            message: error.localizedDescription,
            route: route,
            level: .error,
            traceId: traceId
        )
    }

    /// Capture a structured log event.
    public func captureLog(
        _ message: String,
        level: LogLevel = .info,
        attrs: [String: Any]? = nil,
        traceId: String? = nil,
        release: String? = nil
    ) {
        var attrsJson: String?
        if let attrs = attrs, let data = try? JSONSerialization.data(withJSONObject: attrs) {
            attrsJson = String(data: data, encoding: .utf8)
        }
        let log = LogEvent(
            level: level,
            message: message,
            attrsJson: attrsJson,
            platform: config.platform,
            sessionKey: sessionKey,
            traceId: traceId,
            release: release ?? config.release
        )
        lock.lock(); queuedLogs.append(log); lock.unlock()
        flushIfNeeded()
    }

    /// Capture an API span.
    public func captureSpan(
        method: String,
        route: String,
        status: Int,
        durationMs: Int64,
        traceId: String? = nil,
        spanId: String? = nil,
        release: String? = nil
    ) {
        let span = ApiSpan(
            method: method.uppercased(),
            route: route,
            status: status,
            durationMs: durationMs,
            platform: config.platform,
            traceId: traceId,
            spanId: spanId,
            op: "http.client",
            sessionKey: sessionKey,
            release: release ?? config.release
        )
        lock.lock(); queuedApiSpans.append(span); lock.unlock()
        flushIfNeeded()
    }

    /// Capture a custom RUM event (screen_view, custom vital, etc.)
    public func captureRumEvent(
        type: RumEventType,
        name: String,
        value: Double? = nil,
        route: String? = nil,
        release: String? = nil
    ) {
        let event = RumEvent(
            type: type,
            name: name,
            value: value,
            platform: config.platform,
            route: route,
            sessionKey: sessionKey,
            release: release ?? config.release
        )
        lock.lock(); queuedRumEvents.append(event); lock.unlock()
        flushIfNeeded()
    }

    /// Record a screen view (convenience).
    public func captureScreenView(_ screenName: String) {
        captureRumEvent(type: .screenView, name: screenName, route: screenName)
    }

    // MARK: - Session

    /// Register the current session with the ingest endpoint.
    @discardableResult
    public func startSession(deviceOs: String? = nil, release: String? = nil) -> String {
        let deviceOsStr = deviceOs ?? {
            #if canImport(UIKit)
            return "\(UIDevice.current.systemName) \(UIDevice.current.systemVersion)"
            #else
            return nil
            #endif
        }()
        let session = Session(
            sessionKey: sessionKey,
            platform: config.platform,
            release: release ?? config.release,
            deviceOs: deviceOsStr
        )
        lock.lock(); queuedSessions.append(session); lock.unlock()
        flushIfNeeded()
        return sessionKey
    }

    // MARK: - Direct ingest

    /// Send a fully-formed batch immediately (no buffering).
    @discardableResult
    public func ingest(_ batch: IngestBatch) async throws -> IngestResponse {
        return try await send(batch: batch)
    }

    /// Check the ingest endpoint health.
    public func health() async throws -> HealthResponse {
        guard let url = URL(string: "\(config.ingestUrl)/health") else {
            throw UsermonError.invalidURL
        }
        let (data, resp) = try await urlSession.data(from: url)
        if let http = resp as? HTTPURLResponse, http.statusCode >= 400 {
            throw UsermonError.httpError(statusCode: http.statusCode, body: String(data: data, encoding: .utf8) ?? "")
        }
        return try decoder.decode(HealthResponse.self, from: data)
    }

    // MARK: - Flush

    /// Flush all buffered events immediately.
    @discardableResult
    public func flush() async throws -> IngestResponse? {
        let batch: IngestBatch
        lock.lock()
        let sessions = queuedSessions.isEmpty ? nil : queuedSessions
        let rum = queuedRumEvents.isEmpty ? nil : queuedRumEvents
        let spans = queuedApiSpans.isEmpty ? nil : queuedApiSpans
        let exceptions = queuedExceptions.isEmpty ? nil : queuedExceptions
        let logs = queuedLogs.isEmpty ? nil : queuedLogs
        queuedSessions.removeAll()
        queuedRumEvents.removeAll()
        queuedApiSpans.removeAll()
        queuedExceptions.removeAll()
        queuedLogs.removeAll()
        lock.unlock()

        guard sessions != nil || rum != nil || spans != nil || exceptions != nil || logs != nil else {
            return nil
        }
        batch = IngestBatch(
            sessions: sessions,
            rumEvents: rum,
            apiSpans: spans,
            exceptions: exceptions,
            logs: logs
        )
        return try await send(batch: batch)
    }

    // MARK: - Private

    private func flushIfNeeded() {
        let count: Int
        lock.lock()
        count = queuedSessions.count + queuedRumEvents.count + queuedApiSpans.count
            + queuedExceptions.count + queuedLogs.count
        lock.unlock()
        if count >= config.maxBatchSize {
            Task { try? await flush() }
        }
    }

    private func startFlushTimer() {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.flushTimer = Timer.scheduledTimer(
                withTimeInterval: self.config.flushInterval,
                repeats: true
            ) { [weak self] _ in
                Task { try? await self?.flush() }
            }
        }
    }

    private func send(batch: IngestBatch) async throws -> IngestResponse {
        guard let url = URL(string: "\(config.ingestUrl)/v1/ingest") else {
            throw UsermonError.invalidURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(config.ingestKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        do {
            request.httpBody = try encoder.encode(batch)
        } catch {
            throw UsermonError.encodingError(error)
        }
        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await urlSession.data(for: request)
        } catch {
            throw UsermonError.networkError(error)
        }
        if let http = response as? HTTPURLResponse, http.statusCode >= 400 {
            throw UsermonError.httpError(statusCode: http.statusCode, body: String(data: data, encoding: .utf8) ?? "")
        }
        do {
            return try decoder.decode(IngestResponse.self, from: data)
        } catch {
            throw UsermonError.decodingError(error)
        }
    }
}

// MARK: - Singleton

/// Global singleton — configure once with `Usermon.configure(...)`, then use `Usermon.shared`.
public enum Usermon {
    private static var _shared: UsermonClient?

    /// Configure the global Usermon client.
    ///
    /// Call this once at app launch (e.g. in `application(_:didFinishLaunchingWithOptions:)`).
    @discardableResult
    public static func configure(
        ingestUrl: String,
        ingestKey: String,
        platform: Platform = .ios,
        release: String? = nil,
        tracesSampleRate: Double = 1.0
    ) -> UsermonClient {
        let config = UsermonConfiguration(
            ingestUrl: ingestUrl,
            ingestKey: ingestKey,
            platform: platform,
            release: release,
            tracesSampleRate: tracesSampleRate
        )
        let client = UsermonClient(configuration: config)
        _shared = client
        Task { client.startSession() }
        return client
    }

    /// The configured global client. Crashes if `configure(...)` has not been called.
    public static var shared: UsermonClient {
        guard let client = _shared else {
            fatalError("[Usermon] Call Usermon.configure(...) before accessing Usermon.shared")
        }
        return client
    }

    /// Returns the global client, or nil if not yet configured.
    public static var current: UsermonClient? { _shared }
}
